### ISSUE TITLE
Slow work with hackney_pool

### ISSUE THREAD

**Issue by nateless** [benoitc/hackney#315]
Hey,

I've been trying to move Ruby's code to elixir. The task is simple pass a range 1..1_000_000 to a pool of let's say 200 workers and they should fetch pages in parallel receiving new task as soon as at least some workers are finished.

Here's simplified gist: https://gist.github.com/nateless/2e4cd9ed6171987a640499b23b6c9952

It works, but very slowly, with a pool of 200 workers it's able to do ~100 requests per second, it slowly falls from 180 ops to 100 ops during few hours of work. Ruby's version is constant at ~160 ops.

For test's I've stripped all the work with page data to be able to measure only real http speed.

Is there anything wrong with my code or it's hackney issue. Sorry that it's not erlang, but even for elixir it's really hard for me to shape my mind into functional coding.

**Comment by benoitc** [comment]
@nateless you need to read the body in your tests. Try with the option {with_body, true} to get the body directly in the response or by using the `hackney:body/1` function. Let me know if it doesn't.

That said the new version of hackney that will finally land over the week-end contains a more efficient pooling system ;)

**Comment by nateless** [comment]
It worked thanks :) Have around 500 ops right now. May I ask why reading the body has so profound effect? And what new will be released with next version? :)

**Comment by benoitc** [comment]
@nateless The current design blocks when the maximum of connections is achieved until one is released. Reading the body release the connection at the end and the pool then serve another connection. You can bypass this limit right now by using [hackney_pooler](https://github.com/benoitc/hackney_pooler)

In the coming release hackney won't block anymore and will only keep alive a limited number of connections / host . It will let the application to use its own method to limit the number of concurrent requests like using [jobs](https://github.com/uwiger/jobs) . The advantage of the new system is to remove  possible TCP congestion. However opened connections will still be monitored so if one is idling too long (until you specify if) it will be killed and a timeout will be returned.

**Comment by benoitc** [comment]
updated the comment above.

**Comment by nateless** [comment]
Thank you very much for the information, will read about jobs and will await new hackney's version :) Speed is astonishing comparing to Ruby 👍
